### PR TITLE
Adding a temporary Orphan function to the InstanceDatabase

### PR DESCRIPTION
**Pull request by amzn-tommy**
TEMPOrphan will remove an instance from the database so it will not be found using Find or FindOrCreate. The instance will still persist until its use-count drops to 0, at which point it will be deleted. This is to enable the model asset to remove existing buffer/modellod/model instances and replace them with new instances that have the up to date data. 

Added unit tests for testing. Expanded the parallel tests. In addtion to Creating and immediately letting the refcount drop to 0, I added a mode to hold on to a reference, and added a mode to orphan an instance, and a mode that does both.

Changed the duration for the parallel tests to a float so that small times can be used when needing to run the tests to repro issues locally

Signed-off-by: amzn-tommy <waltont@amazon.com>

### CODE REVIEW

**Review by invertednormal — APPROVED** [review]

**Comment by lawsonamzn** [comment]
This looks like an ok way to deal with it, but doesn't it kind of infer a leak if the asset does _not_ get a  hot reload?  Example, if a "parent" object gets a hot reload, and so it 'orphans' its children, but then only some of them get a hot reload or reload requests (perhaps because a few were already up to date), would this maybe introduce a stall or leak?

**Comment by amzn-tommy** [comment]
> 
> 
> This looks like an ok way to deal with it, but doesn't it kind of infer a leak if the asset does _not_ get a hot reload? Example, if a "parent" object gets a hot reload, and so it 'orphans' its children, but then only some of them get a hot reload or reload requests (perhaps because a few were already up to date), would this maybe introduce a stall or leak?

It depends on how the higher level code is implemented, but I don't believe this can cause a leak in itself. It is analogous to how an existing Asset<T> reference will refer to the old (but still vaild) data, and when OnAssetReloaded is called, you can either keep myOldAssetTReference and it will keep referring to the out of date, but valid data. Or you can set myOldAssetTReference = newReloadedAssetTReference, and the old one will be released and the new one will be used.

The only time you get a leak is if you create a system that will hold on to the old reference, but also hold on to the new reference, so that neither refcount drops to 0.

File: Code/Framework/AtomCore/Tests/InstanceDatabase.cpp [hunk 1]
```diff
@@ -225,23 +253,51 @@ namespace UnitTest
...
-                [&instanceManager, &threadCount, &cv, &guids, &assets, &durationSeconds]()
+                [&instanceManager, &threadCount, &cv, &guids, &instances, &assets, &durationSeconds, &deferRemoval, &orphan, &referenceTableMutex]()
                 {
                     AZ::Debug::Timer timer;
                     timer.Stamp();
 
+                    // If both deferRemoval and orphan are true, use 3 as the modulus so that each iteration will either orphan, release, or create
+                    // If only one is true, use 2 as the modulus so each iteration will create or defer/orphan
+                    // If neither are used, the modulus will be ignored so it doesn't matter
+                    int randomModulus = deferRemoval && orphan ? 3 : 2;
```
Comment (santorac) [inline Code/Framework/AtomCore/Tests/InstanceDatabase.cpp:264]: I found this code to be a bit hard to reason about. Here's a suggested alternative, I'll leave it up to you...
```
bool doCreate = false;
bool doOrphan = false;
bool doDeferredRelease = false;

if(deferRemoval && orphan)
{
    switch(rand() % 3)
    {
        case 0: doCreate = true; break;
        case 1: doOrphan = true; break;
        case 2: doDeferredRelease = true; break;
    }
}
else if(deferRemoval)
{
    switch(rand() % 2)
    {
        case 0: doCreate = true; break;
        case 1: doDeferredRelease = true; break;
    }
}
else if(orphan)
{
    switch(rand() % 2)
    {
        case 0: doCreate = true; break;
        case 1: doOrphan = true; break;
    }
}
else
{
    doCreate = true;
}

if(doOrphan)
{
    instanceManager.TEMPOrphan(instanceId);
}
if(doDeferredRelease)
{
    referenceTableMutex.lock();
    instances[index] = instance;
    referenceTableMutex.unlock();
}
if(doCreate)
{
    Instance<TestInstanceA> instance = instanceManager.FindOrCreate(instanceId, assets[index]);
    ...

    if(deferRemoval)
    {                              
        referenceTableMutex.lock();
        instances[index] = instance;
        referenceTableMutex.unlock();
    }
}
```

File: Code/Framework/AtomCore/Tests/InstanceDatabase.cpp [hunk 2]
```diff
@@ -254,10 +310,12 @@ namespace UnitTest
         // Used to detect a deadlock.  If we wait for more than 10 seconds, it's likely a deadlock has occurred
         while (threadCount > 0 && !timedOut)
         {
+            size_t durationSecondsRoundedUp = static_cast<size_t>(std::ceil(durationSeconds + 1.0f));
```
Comment (santorac) [inline Code/Framework/AtomCore/Tests/InstanceDatabase.cpp:313]: If you want to round up, shouldn't it be "floor(duration+1)" not "ceil(duration+1)"?

**Review by santorac — APPROVED** [review]
I have a couple comments below, but nothing blocking.

File: Code/Framework/AtomCore/Tests/InstanceDatabase.cpp (hunk 2 above)
Comment (amzn-tommy) [reply Code/Framework/AtomCore/Tests/InstanceDatabase.cpp:313]: Should have just been ceil. Will fix.

File: Code/Framework/AtomCore/Tests/InstanceDatabase.cpp (hunk 1 above)
Comment (amzn-tommy) [reply Code/Framework/AtomCore/Tests/InstanceDatabase.cpp:264]: That is definitely more readable. I will make that (or similar) change.

**Review by santorac — APPROVED** [review]

**Pull request merged**